[clang-reorder-fields] Support designated initializers

vvuksanovic opened the pull request:
Initializer lists with designators, missing elements or omitted braces can now be rewritten. Any missing designators are added and they get sorted according to the new order.

```
struct Foo {
  int a;
  int b;
  int c;
};
struct Foo foo = { .a = 1, 2, 3 }
```

when reordering elements to "b,a,c" becomes:

```
struct Foo {
  int b;
  int a;
  int c;
};
struct Foo foo = { .b = 2, .a = 1, .c = 3 }
```

vvuksanovic commented:
@legrosbuffle

legrosbuffle commented on clang-tools-extra/clang-reorder-fields/ReorderFieldsAction.cpp:
```diff
@@ -81,18 +83,52 @@ getNewFieldsOrder(const RecordDecl *Definition,
   return NewFieldsOrder;
 }
 
+struct ReorderedStruct {
```
If you can pull out this refactoring (`ReorderedStruct` + `addRepacement`) to a separate commit that'd be great.

vvuksanovic replied:
Can you clarify what should I pull to a separate commit? I can create a new commit for the `ReorderedStruct` changes, before the designator changes. Should I move `addReplacement` changes there? The new overload would be unused in that commit.

vvuksanovic replied:
Okay, I separated the commits.

legrosbuffle commented on clang-tools-extra/clang-reorder-fields/ReorderFieldsAction.cpp:
```diff
@@ -289,35 +321,213 @@ static void reorderFieldsInConstructor(
 }
 
+/// Replacement for broken InitListExpr::isExplicit function.
+/// TODO: Remove when InitListExpr::isExplicit is fixed.
```
s/TODO/FIXME/

If there and open bug for this ?

legrosbuffle commented on clang-tools-extra/clang-reorder-fields/ReorderFieldsAction.cpp:
```diff
@@ -289,35 +321,213 @@ static void reorderFieldsInConstructor(
+/// they are equal.
+static int cmpDesignators(const DesignatorIter &Lhs, const DesignatorIter &Rhs,
+                          const ReorderedStruct &Struct) {
+  if (Lhs.getTag() == DesignatorIter::STRUCT) {
```
a `switch` might make this a bit more readable

vvuksanovic replied:
Done.

legrosbuffle reviewed (commented):
Nice ! Conceptually this looks good, only stylistic comments.

vvuksanovic commented on clang-tools-extra/clang-reorder-fields/utils/Designator.cpp:
```diff
@@ -0,0 +1,256 @@
+namespace clang {
+namespace reorder_fields {
+
+DesignatorIter::DesignatorIter(const QualType Type,
```
Done.

vvuksanovic commented:
ping @legrosbuffle

vvuksanovic committed:
[clang-reorder-fields] Refactor the new order information

Move the new fields order information to a new struct.

vvuksanovic committed:
[clang-reorder-fields] Support designated initializers

Initializer lists with designators, missing elements or omitted braces can now be rewritten. Any missing designators are added and they get sorted according to the new order.

```
struct Foo {
int a;
int b;
int c;
};
struct Foo foo = { .a = 1, 2, 3 }
```

when reordering elements to "b,a,c" becomes:

```
struct Foo {
int b;
int a;
int c;
};
struct Foo foo = { .b = 2, .a = 1, .c = 3 }
```

vvuksanovic committed:
fixup! [clang-reorder-fields] Support designated initializers

vvuksanovic committed:
fixup! [clang-reorder-fields] Support designated initializers

alexander-shaposhnikov commented:
I've added a couple of comments, but overall this looks reasonable to me, thank your for your patience

alexander-shaposhnikov reviewed (approved):
LGTM (modulo comments)

vvuksanovic committed:
fixup! [clang-reorder-fields] Support designated initializers

vvuksanovic commented on clang-tools-extra/test/clang-reorder-fields/DesignatedInitializerList.c:
Done. Added `DesignatedInitializerList.cpp` and used the existing `AggregatePartialInitialization.cpp` test to check for the unsupported message.

There is a slight difference in implementation now. I found that C++20 designators actually work in clang and gcc even with older standards, so we now add designators if some fields are already designated, but we issue a warning otherwise (unless using C++20).

vvuksanovic commented on clang-tools-extra/clang-reorder-fields/Designator.cpp:
```diff
@@ -0,0 +1,213 @@
+  for (auto &I : DesignatorList) {
+    switch (I.getTag()) {
+    case DesignatorIter::STRUCT:
+      Designator += "." + I.getStructIter()->getName().str();
```
Used `raw_string_ostream` instead.

vvuksanovic commented on clang-tools-extra/test/clang-reorder-fields/DesignatedInitializerList.c:
Added `InitializerListExcessElements.c` and `IdiomaticZeroInitializer.c` tests. I can't find test cases for non constant-sized arrays in structs. I would change those checks for asserts, but there might be some way to trigger that, that I didn't think of.

Also modified the message to include the location of the problematic initializer list.

alexander-shaposhnikov commented:
lgtm

vvuksanovic commented:
Thank you for your patience. Can you please merge?

vvuksanovic commented:
Thanks!
